035-ca-abstraction.md

```diff
@@ -0,0 +1,396 @@
+
+4. Once a certificate is issued, the issuing operator needs check whether the CA certificate is trusted. 
+   I.e. it needs to know what trust states all the existing CA certificates are in. 
+   If it is not trusted **the issuing operator puts the end entity in the `PENDING` state.** and **puts the CA certificate in the `UNTRUSTED` state**. (**See step 3 in the diagram.**)
+
+> From the above it is clear that there is a `Secret` which stores tokens and issued certificates and their current state, and that only the issuing operator needs access to it.
+> It's also clear that there is a `Secret` which stores CA certificates, and their current state, and which needs to be accessed by the issuing operator and the cluster operator.
+
+#### Process for ensuring CA trust
+
+When an EE certificate is issued via a `CertificateIssuer` it might have been signed by a CA that is not yet trusted.
+The cluster operator (for the cluster CA), or the user operator (for the clients CA) needs to detect this lack of established trust and ask the cluster operator to trust the new CA. 
+The cluster operator and the user operator run in separate processes, so this will be done via a shared `Secret`, for example:
+
+2. The issuing operator has placed a root CA certificate in the shared `Secret` in the `UNTRUSTED` state.
+3. The cluster operator sees the root CA certificate orchestrates changes to the brokers (reconfig or restart) (and, for the Cluster CA the other components) so that the certificate is trusted by all components that will observe the EE certificates.
+4. The cluster operator updates the shared `Secret` to indicate that the certificate is now in the `TRUSTED_NEW` state.
+
+#### Process for deploying the issued certificate
+
+1. The issuing operator sees that the root CA certificate is now in the `TRUSTED_NEW` state, so can infer that the CA certificate is trusted.
+2. The issuing operator copies the EE certificate and key, and the rest of the certificate chain from its `Secret` _B_  to the deployment secret (e.g. the "user secret" for the Clients CA, or the relevant component secret for the Cluster CA).
+3. The issuing operator updates the state of the root CA certificate in the shared `Secret` to `TRUSTED_USED`. 
+
+#### Need for cert path validation
+
+The issuing operator needs to know when to request a new certificate for the same subject because the existing certificate is (or is soon going to be) invalid.
+Validity is a property of the whole certificate chain, not just the EE certificate.
+For example it is possible that a root or intermediate certificate expires before an EE certificate that depends on it. 
+So for the operator to reason correctly about validity it needs to perform PKIX cert path validation.
+Fortunately Java comes with APIs for doing this in `java.security.cert`.
```

Comment:
It's very configurable. You can configure it with your own effective date and it will give you detailed results about which cert is invalid and why. See [this guide](https://docs.oracle.com/javase/7/docs/technotes/guides/security/certpath/CertPathProgGuide.html) and the [API docs](https://docs.oracle.com/javase/7/docs/api/java/security/cert/package-summary.html).